### PR TITLE
Refactoring classes to solve the issue #27

### PR DESCRIPTION
**Pull request by jrdalpra**
Solving issue https://github.com/jcabi/jcabi-ssh/issues/27

### CODE REVIEW

**Comment by jrdalpra** [comment]
@dmarkov can you assign someone to review this PR? Thanks.

**Comment by dmarkov** [comment]
@jrdalpra Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by dmarkov** [comment]
@dmzaytsev it's yours, please review

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import org.apache.commons.lang3.Validate;
+
+/**
+ * Common module for any ssh shell.
+ * @author Jose Volmei Dal Pra Junior (jrdalpra@gmail.com)
+ * @version $Id$
+ * @since 2.0
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @jrdalpra  `1.5.2` ?

latest release is `1.5.1` https://github.com/jcabi/jcabi-ssh/releases

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    private final transient String login;
+
+    /**
+     * Constructor.
+     * @param adr Address that you want to connect to.
+     * @param prt Port that you want to reach.
+     * @param user User that will be used when connecting.
+     * @throws UnknownHostException when host is unkwon.
+     */
+    public AbstractSSHShell(
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @jrdalpra why do you need public constructor in this abstract class?
I think package level access will be enough for this one

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        Validate.matchesPattern(
+            this.addr,
+            "\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}",
+            "Invalid IP address of the server `%s`",
+            this.addr
+        );
+        Validate.notEmpty(user, "user name can't be empty");
+    }
+
+    // @checkstyle ParameterNumberCheck (6 lines)
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @jrdalpra  2 line enough here

File: src/main/java/com/jcabi/ssh/SSH.java [hunk 4]
```diff
@@ -67,18 +64,15 @@
...
  * @version $Id$
  * @since 1.0
  * @see <a href="http://www.yegor256.com/2014/09/02/java-ssh-client.html">article by Yegor Bugayenko</a>
  */
 @ToString
-@EqualsAndHashCode(of = { "addr", "port", "login", "key" })
+@EqualsAndHashCode(of = { "key" }, callSuper = true)
 @SuppressWarnings("PMD.TooManyMethods")
-public final class SSH implements Shell {
+public final class SSH extends AbstractSSHShell implements Shell {
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/SSH.java]: @jrdalpra `Shell` interface declaration is redundant here

File: src/main/java/com/jcabi/ssh/SSH.java [hunk 5]
```diff
@@ -243,7 +199,7 @@ public static String escape(final String arg) {
         randomize = true,
         types = IOException.class
     )
-    private Session session() throws IOException {
+    public Session session() throws IOException {
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/SSH.java]: @jrdalpra why did you change the access level of this method?

File: src/main/java/com/jcabi/ssh/SSHByPassword.java [hunk 6]
```diff
@@ -53,23 +49,8 @@
...
-     * Port to use.
-     */
-    private final transient int port;
-
-    /**
-     * User name.
-     */
-    private final transient String login;
+@EqualsAndHashCode(of = { "password" }, callSuper = true)
+public final class SSHByPassword extends AbstractSSHShell implements Shell {
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/SSHByPassword.java]: @jrdalpra `Shell` interface declaration is redundant here

File: src/main/java/com/jcabi/ssh/SSHByPassword.java [hunk 7]
```diff
@@ -124,7 +82,7 @@ public int exec(final String command, final InputStream stdin,
         randomize = true,
         types = IOException.class
     )
-    private Session session() throws IOException {
+    public Session session() throws IOException {
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/SSHByPassword.java]: @jrdalpra why did you change the access level of this method?

**Comment by dmzaytsev** [comment]
@jrdalpra please see a few comments above

File: src/main/java/com/jcabi/ssh/SSHByPassword.java (hunk 7 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/SSHByPassword.java]: @dmzaytsev qulice is complaing about it

Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/SSHByPassword.java]: @dmzaytsev it was transfered to AbstractSSHShell. So they need to me protected or public.

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java (hunk 1 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @dmzaytsev done

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java (hunk 2 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @dmzaytsev done

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java (hunk 3 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @dmzaytsev sorry, I didn't get id

File: src/main/java/com/jcabi/ssh/SSH.java (hunk 4 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/SSH.java]: @dmzaytsev done

File: src/main/java/com/jcabi/ssh/SSH.java (hunk 5 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/SSH.java]: @dmzaytsev done. Now it has package level

File: src/main/java/com/jcabi/ssh/SSHByPassword.java (hunk 6 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/SSHByPassword.java]: @dmzaytsev done

File: src/main/java/com/jcabi/ssh/SSHByPassword.java (hunk 7 above)
Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/SSHByPassword.java]: @dmzaytsev done. Package level now.

**Comment by jrdalpra** [comment]
@dmzaytsev done

**Comment by jrdalpra** [comment]
@dmzaytsev I'm getting this error "Use explicit scoping instead of the default package private level (DefaultPackage)"

**Comment by jrdalpra** [comment]
@dmzaytsev I ignored the checkstyle. I don't think it's a big problem to have protected method. Do you think?

File: src/main/java/com/jcabi/ssh/AbstractSSHShell.java (hunk 3 above)
Comment (dmzaytsev) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @jrdalpra you can use 
`@checkstyle ParameterNumberCheck (2 lines)` 
instead of 
`@checkstyle ParameterNumberCheck (6 lines)`

Comment (jrdalpra) [inline src/main/java/com/jcabi/ssh/AbstractSSHShell.java]: @dmzaytsev done

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by jrdalpra** [comment]
@yegor256 can you deploy a new version?

**Comment by yegor256** [comment]
@rultor release, tag is `1.5.2`

**Comment by dmarkov** [comment]
@dmzaytsev Thanks so much! Your account was topped up for **25 mins** (transaction ID is `73938882`, the task took 119 hours and 15 mins)

you're getting extra minutes here (c=10)

added +25 to your rating, now it is equal to [+1225](http://www.netbout.com/b/35166?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy pls
